src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaTopicScaler.cs

```diff
@@ -0,0 +1,249 @@
+
+                    if (diff > highestPartitionLag)
+                    {
+                        highestPartitionLag = diff;
+                        partitionWithHighestLag = topicPartition.Partition;
+                    }
+                }
+            }
+
+            if (partitionWithHighestLag != Partition.Any)
+            {
+                logger.LogInformation("Total lag in '{topic}' is {totalLag}, highest partition lag found in {partition} with value of {offsetDifference}", this.topicName, totalLag, partitionWithHighestLag.Value, highestPartitionLag);
+            }
+
+            return Task.FromResult(new KafkaTriggerMetrics(totalLag, allPartitions.Count));
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.OfType<KafkaTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<KafkaTriggerMetrics> context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, KafkaTriggerMetrics[] metrics)
+        {
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None,
+            };
+
+            const int NumberOfSamplesToConsider = 5;
+
+            // At least 5 samples are required to make a scale decision for the rest of the checks.
+            if (metrics == null || metrics.Length < NumberOfSamplesToConsider)
+            {
+                return status;
+            }
+
+            var lastMetrics = metrics.Last();
+            long totalLag = lastMetrics.TotalLag;
+            long partitionCount = lastMetrics.PartitionCount;
+            long lagThreshold = 1000L;
+
+            //This case shouldn't ever happens as we do not scale out in case workerCount >= partitionCount
+            if (partitionCount > 0 && partitionCount < workerCount)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+
+                if (this.logger.IsEnabled(LogLevel.Information))
+                {
+                    this.logger.LogInformation("WorkerCount ({workerCount}) > PartitionCount ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                    this.logger.LogInformation("Number of instances ({workerCount}) is too high relative to number of partitions ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                }
+
+                return status;
+            }
+
+
+            // Check to see if the Kafka consumer has been empty for a while. Only if all metrics samples are empty do we scale down.
+            bool partitionIsIdle = metrics.All(p => p.TotalLag == 0);
```

Comment:
Do we want to check that all metrics have TotalLag == 0 or just the last 5?

Reply:
Note, this will be the metrics in the last 2 min

Reply:
Since I don't know how this method will be called I am guessing only. If we do get a list containing hours of metrics and at least of them had lag we might incorrectly assume that there is lag (i.e. we had lag 1 hour ago, but the last 5 metrics indicate no lag).

Reply:
In that case, knowing it will be 2 minutes metrics it looks good to me.

Reply:
Sounds I will leave it for @mathewc to guide.

Reply:
I'd just check all as you've done. That's what the other extensions do.

Reply:
is it worth making this 2mins configurable, or not?

Reply:
It is already configurable to a degree - see [ScaleOptions](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Config/ScaleOptions.cs) in the Functions host. The host is the one taking metrics samples, and the number of samples will be governed by the sample interval.